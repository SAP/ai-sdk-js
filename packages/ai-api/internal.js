export * from './dist/internal.js';
//# sourceMappingURL=internal.js.map
