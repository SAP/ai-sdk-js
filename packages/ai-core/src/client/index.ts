// This file is only needed for the public API check
// It will be overwritten when regenerating, we will need a better solution soon
// Also, if there is nothing in this file, the CJS check fails, therefore the dummy
export const dummy = 'dummy';
