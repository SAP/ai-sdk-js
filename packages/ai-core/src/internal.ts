export * from './client/index.js';
export * from './utils/index.js';
