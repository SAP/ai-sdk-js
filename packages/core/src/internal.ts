export * from './context.js';
export * from './http-client.js';
export * from './openapi-request-builder.js';
export * from './model-types.js';
