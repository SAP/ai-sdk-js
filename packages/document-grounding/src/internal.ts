export * from './client/api/index.js';
