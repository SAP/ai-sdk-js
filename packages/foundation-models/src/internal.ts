export * from './azure-openai/index.js';
