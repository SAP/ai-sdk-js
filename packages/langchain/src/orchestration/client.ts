import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { OrchestrationClient as OrchestrationClientBase } from '@sap-ai-sdk/orchestration';
import {
  isTemplate,
  mapLangchainMessagesToOrchestrationMessages,
  mapOutputToChatResult
} from './util.js';
import type { BaseLanguageModelInput } from '@langchain/core/language_models/base';
import type { Runnable, RunnableLike } from '@langchain/core/runnables';
import type { OrchestrationMessageChunk } from './orchestration-message-chunk.js';
import type { ChatResult } from '@langchain/core/outputs';
import type { OrchestrationModuleConfig } from '@sap-ai-sdk/orchestration';
import type { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import type { ResourceGroupConfig } from '@sap-ai-sdk/ai-api';
import type { BaseMessage } from '@langchain/core/messages';
import type { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { OrchestrationCallOptions } from './types.js';
import type { HttpDestinationOrFetchOptions } from '@sap-cloud-sdk/connectivity';

/**
 * The Orchestration client.
 */
export class OrchestrationClient extends BaseChatModel<
  OrchestrationCallOptions,
  OrchestrationMessageChunk
> {
  constructor(
    // TODO: Omit streaming until supported
    public orchestrationConfig: Omit<OrchestrationModuleConfig, 'streaming'>,
    public langchainOptions: BaseChatModelParams = {},
    public deploymentConfig?: ResourceGroupConfig,
    public destination?: HttpDestinationOrFetchOptions
  ) {
    super(langchainOptions);
  }

  _llmType(): string {
    return 'orchestration';
  }

  /**
   * Create a new runnable sequence that runs each individual runnable in series,
   * piping the output of one runnable into another runnable or runnable-like.
   * @param coerceable - A runnable, function, or object whose values are functions or runnables.
   * @returns A new runnable sequence.
   */
  override pipe<NewRunOutput>(
    coerceable: RunnableLike<OrchestrationMessageChunk, NewRunOutput>
  ): Runnable<
    BaseLanguageModelInput,
    Exclude<NewRunOutput, Error>,
    OrchestrationCallOptions
  > {
    return super.pipe(coerceable) as Runnable<
      BaseLanguageModelInput,
      Exclude<NewRunOutput, Error>,
      OrchestrationCallOptions
    >;
  }

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal
      },
      () => {
        const { inputParams, customRequestConfig } = options;
        const mergedOrchestrationConfig =
          this.mergeOrchestrationConfig(options);
        const orchestrationClient = new OrchestrationClientBase(
          mergedOrchestrationConfig,
          this.deploymentConfig,
          this.destination
        );
        const messagesHistory =
          mapLangchainMessagesToOrchestrationMessages(messages);
        return orchestrationClient.chatCompletion(
          {
            messagesHistory,
            inputParams
          },
          customRequestConfig
        );
      }
    );

    const content = res.getContent();

    // TODO: Add streaming as soon as we support it
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }

  private mergeOrchestrationConfig(
    options: typeof this.ParsedCallOptions
  ): OrchestrationModuleConfig {
    const { tools = [], stop = [] } = options;
    return {
      ...this.orchestrationConfig,
      llm: {
        ...this.orchestrationConfig.llm,
        model_params: {
          ...this.orchestrationConfig.llm.model_params,
          ...(stop.length && {
            stop: [
              ...(this.orchestrationConfig.llm.model_params?.stop || []),
              ...stop
            ]
          })
        }
      },
      templating: {
        ...this.orchestrationConfig.templating,
        ...(isTemplate(this.orchestrationConfig.templating) &&
          tools.length && {
            tools: [
              ...(this.orchestrationConfig.templating.tools || []),
              ...tools
            ]
          })
      }
    };
  }
}
