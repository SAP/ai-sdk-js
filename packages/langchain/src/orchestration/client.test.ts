import { constructCompletionPostRequest } from '@sap-ai-sdk/orchestration/internal.js';
import { jest } from '@jest/globals';
import nock from 'nock';
import { type AIMessageChunk } from '@langchain/core/messages';
import {
  mockClientCredentialsGrantCall,
  mockDeploymentsList,
  mockInference,
  parseMockResponse,
  parseFileToString
} from '../../../../test-util/mock-http.js';
import { addNumbersTool } from '../../../../test-util/tools.js';
import { OrchestrationClient } from './client.js';
import type { LangChainOrchestrationModuleConfig } from './types.js';
import type { ToolCall } from '@langchain/core/messages/tool';
import type {
  CompletionPostResponse,
  ErrorResponse
} from '@sap-ai-sdk/orchestration';

jest.setTimeout(30000);

describe('orchestration service client', () => {
  let mockResponse: CompletionPostResponse;
  let mockResponseInputFilterError: ErrorResponse;
  let mockResponseStream: string;
  let mockResponseStreamToolCalls: string;
  const messages = [{ role: 'user' as const, content: 'Hello!' }];
  const config: LangChainOrchestrationModuleConfig = {
    llm: {
      model_name: 'gpt-4o',
      model_params: {}
    }
  };
  const endpoint = {
    url: 'inference/deployments/1234/completion'
  };

  beforeAll(async () => {
    mockClientCredentialsGrantCall();
    mockDeploymentsList({ scenarioId: 'orchestration' }, { id: '1234' });
    mockResponse = await parseMockResponse<CompletionPostResponse>(
      'orchestration',
      'orchestration-chat-completion-success-response.json'
    );
    mockResponseInputFilterError = await parseMockResponse<ErrorResponse>(
      'orchestration',
      'orchestration-chat-completion-input-filter-error.json'
    );
    mockResponseStream = await parseFileToString(
      'orchestration',
      'orchestration-chat-completion-stream-chunks.txt'
    );
    mockResponseStreamToolCalls = await parseFileToString(
      'orchestration',
      'orchestration-chat-completion-stream-tools-chunks.txt'
    );
  });

  afterEach(() => {
    nock.cleanAll();
  });

  function mockInferenceWithResilience(
    response: any,
    resilience: {
      retry?: number;
      delay?: number;
    },
    status: number = 200,
    isStream?: boolean
  ) {
    mockInference(
      {
        data: constructCompletionPostRequest(
          {
            ...config,
            templating: {
              template: messages
            }
          },
          { messages: [] },
          isStream
        )
      },
      {
        data: response,
        status
      },
      endpoint,
      resilience
    );
  }

  describe('resilience', () => {
    it('returns successful response when maxRetries equals retry configuration', async () => {
      mockInferenceWithResilience(mockResponse, { retry: 2 });
      const client = new OrchestrationClient(config, {
        maxRetries: 2
      });
      expect(await client.invoke(messages)).toMatchSnapshot();
    });

    it('throws error response when maxRetries is smaller than required retries', async () => {
      mockInferenceWithResilience(mockResponse, { retry: 2 });
      const client = new OrchestrationClient(config, {
        maxRetries: 1
      });
      await expect(client.invoke(messages)).rejects.toThrow(
        'Request failed with status code 500'
      );
    });

    it('throws when delay exceeds timeout', async () => {
      mockInferenceWithResilience(mockResponse, { delay: 2000 });
      const client = new OrchestrationClient(config);
      const response = client.invoke(messages, { timeout: 1000 });
      await expect(response).rejects.toThrow(
        expect.objectContaining({
          stack: expect.stringMatching(/Timeout/)
        })
      );
    });

    it('throws immediately when input filter error occurs', async () => {
      mockInferenceWithResilience(
        mockResponseInputFilterError,
        { retry: 0 },
        400
      );
      const client = new OrchestrationClient(config, {
        maxRetries: 1000 // Retry forever unless input filter error
      });
      await expect(client.invoke(messages)).rejects.toThrow(
        'Request failed with status code 400'
      );
    }, 1000);

    it('throws when delay exceeds timeout during streaming', async () => {
      mockInferenceWithResilience(
        mockResponseStream,
        { delay: 2000 },
        200,
        true
      );

      let finalOutput: AIMessageChunk | undefined;
      const client = new OrchestrationClient(config, { maxRetries: 0 });
      try {
        const stream = await client.stream([], { timeout: 1000 });
        for await (const chunk of stream) {
          finalOutput = finalOutput ? finalOutput.concat(chunk) : chunk;
        }
      } catch (e) {
        expect(e).toEqual(
          expect.objectContaining({
            stack: expect.stringMatching(/Timeout/)
          })
        );
      }
    });

    it('returns successful response when timeout is bigger than delay', async () => {
      mockInferenceWithResilience(mockResponse, { delay: 2000 });
      const client = new OrchestrationClient(config);
      const response = await client.invoke(messages, { timeout: 3000 });
      expect(response).toMatchSnapshot();
    });

    it('throws immediately when input filter error occurs', async () => {
      mockInferenceWithResilience(
        mockResponseInputFilterError,
        { retry: 0 },
        400
      );
      const client = new OrchestrationClient(config, {
        maxRetries: 1000 // Retry forever unless input filter error
      });
      await expect(client.invoke(messages)).rejects.toThrow(
        'Request failed with status code 400'
      );
    }, 1000);

    it('throws when delay exceeds timeout using streaming', async () => {
      mockInferenceWithResilience(
        mockResponseStream,
        { delay: 2000 },
        200,
        true
      );
      const client = new OrchestrationClient(config, { maxRetries: 0 });
      await expect(client.stream('Hello!', { timeout: 1000 })).rejects.toThrow(
        'AbortError'
      );
    });
  });

  describe('bindTools', () => {
    let client: OrchestrationClient;
    const toolResponse = {
      data: {
        orchestration_result: {
          choices: [
            {
              message: {
                role: 'assistant',
                content: null,
                tool_calls: [
                  {
                    id: '1',
                    type: 'function',
                    function: {
                      name: 'add',
                      arguments: '{ "a": 1, "b": 2 }'
                    }
                  }
                ]
              },
              index: 0
            }
          ]
        }
      }
    };

    beforeEach(() => {
      client = new OrchestrationClient(config, { maxRetries: 0 });
    });

    it('should bind a tool with strict set to true if defined in kwargs', async () => {
      mockInference(
        {
          data: {
            orchestration_config: {
              module_configurations: {
                templating_module_config: {
                  tools: [
                    {
                      type: 'function',
                      function: {
                        ...addNumbersTool.function,
                        strict: true // Will be tested
                      }
                    }
                  ],
                  template: [{ role: 'user', content: 'What is 1 + 2?' }]
                },
                llm_module_config: {
                  model_name: 'gpt-4o',
                  model_params: {}
                }
              }
            }
          }
        },
        toolResponse,
        endpoint
      );
      await client
        .bindTools([addNumbersTool], { strict: true })
        .invoke('What is 1 + 2?');
    });

    it('should bind a tool with strict set to false if defined in kwargs', async () => {
      mockInference(
        {
          data: {
            orchestration_config: {
              module_configurations: {
                templating_module_config: {
                  tools: [
                    {
                      type: 'function',
                      function: {
                        ...addNumbersTool.function,
                        strict: false // Will be tested
                      }
                    }
                  ],
                  template: [{ role: 'user', content: 'What is 1 + 2?' }]
                },
                llm_module_config: {
                  model_name: 'gpt-4o',
                  model_params: {}
                }
              }
            }
          }
        },
        toolResponse,
        endpoint
      );
      await client
        .bindTools([addNumbersTool], { strict: false })
        .invoke('What is 1 + 2?');
    });

    it('should bind a tool with undefined strict if not defined in kwargs', async () => {
      mockInference(
        {
          data: {
            orchestration_config: {
              module_configurations: {
                templating_module_config: {
                  tools: [
                    {
                      type: 'function',
                      function: {
                        ...addNumbersTool.function,
                        strict: undefined // Will be tested
                      }
                    }
                  ],
                  template: [{ role: 'user', content: 'What is 1 + 2?' }]
                },
                llm_module_config: {
                  model_name: 'gpt-4o',
                  model_params: {}
                }
              }
            }
          }
        },
        toolResponse,
        endpoint
      );
      await client.bindTools([addNumbersTool]).invoke('What is 1 + 2?');
    });
  });

  describe('streaming', () => {
    it('supports streaming responses', async () => {
      mockInference(
        {
          data: constructCompletionPostRequest(
            {
              ...config,
              templating: {
                template: messages
              }
            },
            { messages: [] },
            true
          )
        },
        {
          data: mockResponseStream,
          status: 200
        },
        endpoint
      );

      const client = new OrchestrationClient(config);
      const stream = await client.stream('Hello!');
      let finalOutput: AIMessageChunk | undefined;

      for await (const chunk of stream) {
        finalOutput = finalOutput ? finalOutput.concat(chunk) : chunk;
      }
      expect(finalOutput).toMatchSnapshot();
    });

    it('streams and aborts with a signal', async () => {
      mockInference(
        {
          data: constructCompletionPostRequest(
            {
              ...config,
              templating: {
                template: messages
              }
            },
            { messages: [] },
            true
          )
        },
        {
          data: mockResponseStream,
          status: 200
        },
        endpoint
      );
      const client = new OrchestrationClient(config);
      const controller = new AbortController();
      const { signal } = controller;
      const stream = await client.stream('Hello!', { signal });
      const streamFunction = async () => {
        for await (const _ of stream) {
          controller.abort();
        }
      };

      await expect(streamFunction()).rejects.toThrow('Aborted');
    }, 1000);

    it('streams with a callback', async () => {
      mockInference(
        {
          data: constructCompletionPostRequest(
            {
              ...config,
              templating: {
                template: messages
              }
            },
            { messages: [] },
            true
          )
        },
        {
          data: mockResponseStream,
          status: 200
        },
        endpoint
      );
      let tokenCount = 0;
      const callbackHandler = {
        handleLLMNewToken: jest.fn().mockImplementation(() => {
          tokenCount += 1;
        })
      };
      const client = new OrchestrationClient(config, {
        callbacks: [callbackHandler]
      });
      const stream = await client.stream('Hello!');
      const chunks: AIMessageChunk[] = [];

      for await (const chunk of stream) {
        chunks.push(chunk);
        break;
      }
      expect(callbackHandler.handleLLMNewToken).toHaveBeenCalled();
      const firstCallArgs = callbackHandler.handleLLMNewToken.mock.calls[0];
      // First chunk content is empty
      expect(firstCallArgs[0]).toEqual('');
      // Second argument should be the token indices
      expect(firstCallArgs[1]).toEqual({ prompt: 0, completion: 0 });
      expect(tokenCount).toBeGreaterThan(0);
    });

    it('supports streaming responses with tool calls', async () => {
      mockInference(
        {
          data: constructCompletionPostRequest(
            {
              ...config,
              templating: {
                template: messages
              }
            },
            { messages: [] },
            true
          )
        },
        {
          data: mockResponseStreamToolCalls,
          status: 200
        },
        endpoint
      );

      const client = new OrchestrationClient(config);
      const stream = await client.stream('Hello!');

      let finalOutput: AIMessageChunk | undefined;
      for await (const chunk of stream) {
        finalOutput = finalOutput ? finalOutput.concat(chunk) : chunk;
      }
      const completeToolCall: ToolCall = finalOutput!.tool_calls![0];
      expect(completeToolCall.name).toEqual('add');
      expect(completeToolCall.args).toEqual({
        a: 2,
        b: 3
      });
      expect(finalOutput).toMatchSnapshot();
    });
  });
});
