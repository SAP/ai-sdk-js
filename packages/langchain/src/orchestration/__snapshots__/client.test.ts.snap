// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orchestration service client resilience returns successful response when maxRetries equals retry configuration 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessage",
  ],
  "kwargs": {
    "additional_kwargs": {
      "finish_reason": "stop",
      "function_call": undefined,
      "index": 0,
      "module_results": {
        "llm": {
          "choices": [
            {
              "finish_reason": "stop",
              "index": 0,
              "message": {
                "content": "Hello! How can I assist you today?",
                "role": "assistant",
              },
            },
          ],
          "created": 172,
          "id": "llm-id",
          "model": "gpt-4o",
          "object": "chat.completion",
          "usage": {
            "completion_tokens": 9,
            "prompt_tokens": 9,
            "total_tokens": 18,
          },
        },
        "templating": [
          {
            "content": "Hello!",
            "role": "user",
          },
        ],
      },
      "request_id": "request-id",
      "tool_calls": undefined,
    },
    "content": "Hello! How can I assist you today?",
    "invalid_tool_calls": [],
    "response_metadata": {
      "created": 172,
      "finish_reason": "stop",
      "function_call": undefined,
      "id": "orchestration-id",
      "index": 0,
      "model": "gpt-4o",
      "object": "chat.completion",
      "system_fingerprint": undefined,
      "tokenUsage": {
        "completionTokens": 9,
        "promptTokens": 9,
        "totalTokens": 18,
      },
      "tool_calls": undefined,
    },
    "tool_calls": [],
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`orchestration service client resilience returns successful response when timeout is bigger than delay 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessage",
  ],
  "kwargs": {
    "additional_kwargs": {
      "finish_reason": "stop",
      "function_call": undefined,
      "index": 0,
      "module_results": {
        "llm": {
          "choices": [
            {
              "finish_reason": "stop",
              "index": 0,
              "message": {
                "content": "Hello! How can I assist you today?",
                "role": "assistant",
              },
            },
          ],
          "created": 172,
          "id": "llm-id",
          "model": "gpt-4o",
          "object": "chat.completion",
          "usage": {
            "completion_tokens": 9,
            "prompt_tokens": 9,
            "total_tokens": 18,
          },
        },
        "templating": [
          {
            "content": "Hello!",
            "role": "user",
          },
        ],
      },
      "request_id": "request-id",
      "tool_calls": undefined,
    },
    "content": "Hello! How can I assist you today?",
    "invalid_tool_calls": [],
    "response_metadata": {
      "created": 172,
      "finish_reason": "stop",
      "function_call": undefined,
      "id": "orchestration-id",
      "index": 0,
      "model": "gpt-4o",
      "object": "chat.completion",
      "system_fingerprint": undefined,
      "tokenUsage": {
        "completionTokens": 9,
        "promptTokens": 9,
        "totalTokens": 18,
      },
      "tool_calls": undefined,
    },
    "tool_calls": [],
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`orchestration service client streaming supports streaming responses 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {
      "module_results": {
        "llm": {
          "choices": [
            {
              "delta": {
                "content": "The SAP Cloud SDK is a comprehensive development toolkit designed to simplify and accelerate the creation of applications that integrate with SAP solutions, particularly those built on the SAP Business Technology Platform (BTP). It provides developers with libraries, tools, and best practices that streamline the process of connecting to SAP systems, such as S/4HANA and other services available on the SAP Cloud Platform.

Key features of the SAP Cloud SDK include:

1. **Simplified Connectivity**: The SDK offers pre-built libraries to easily interact with SAP services, providing capabilities for authentication, service consumption, and OData/REST client generation.

2. **Multi-cloud Support**: It supports multiple cloud environments, ensuring that applications remain flexible and can be deployed across various cloud providers.

3. **Best Practices and Guidelines**: The SDK includes best practices for development, ensuring high-quality, scalable, and maintainable code.

4. **Project Scaffolding and Code Samples**: Developers can quickly start their projects using provided templates and samples, accelerating the development process and reducing the learning curve.

5. **Extensive Documentation and Community Support**: Ample documentation, tutorials, and an active community help developers overcome challenges and adopt the SDK efficiently.

Overall, the SAP Cloud SDK is an essential tool for developers looking to build cloud-native applications and extensions that seamlessly integrate with SAP's enterprise solutions.",
                "role": "assistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistant",
              },
              "finish_reason": "stop",
              "index": 0,
            },
          ],
          "created": 1734524005,
          "id": "chatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wpchatcmpl-AfnDZfYvuE4SDplaLGF9v0PJjB0wp",
          "model": "gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06",
          "object": "chat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunk",
          "system_fingerprint": "fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48fp_4e924a4b48",
          "usage": {
            "completion_tokens": 271,
            "prompt_tokens": 17,
            "total_tokens": 288,
          },
        },
        "templating": [
          {
            "content": "Give me a short introduction of SAP Cloud SDK.",
            "role": "user",
          },
        ],
      },
      "request_id": "66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b66172762-8c47-4438-89e7-2689be8f370b",
    },
    "content": "The SAP Cloud SDK is a comprehensive development toolkit designed to simplify and accelerate the creation of applications that integrate with SAP solutions, particularly those built on the SAP Business Technology Platform (BTP). It provides developers with libraries, tools, and best practices that streamline the process of connecting to SAP systems, such as S/4HANA and other services available on the SAP Cloud Platform.

Key features of the SAP Cloud SDK include:

1. **Simplified Connectivity**: The SDK offers pre-built libraries to easily interact with SAP services, providing capabilities for authentication, service consumption, and OData/REST client generation.

2. **Multi-cloud Support**: It supports multiple cloud environments, ensuring that applications remain flexible and can be deployed across various cloud providers.

3. **Best Practices and Guidelines**: The SDK includes best practices for development, ensuring high-quality, scalable, and maintainable code.

4. **Project Scaffolding and Code Samples**: Developers can quickly start their projects using provided templates and samples, accelerating the development process and reducing the learning curve.

5. **Extensive Documentation and Community Support**: Ample documentation, tutorials, and an active community help developers overcome challenges and adopt the SDK efficiently.

Overall, the SAP Cloud SDK is an essential tool for developers looking to build cloud-native applications and extensions that seamlessly integrate with SAP's enterprise solutions.",
    "id": undefined,
    "invalid_tool_calls": [],
    "response_metadata": {
      "completion": 0,
      "finish_reason": "stop",
      "prompt": 0,
      "token_usage": {
        "completion_tokens": 271,
        "prompt_tokens": 17,
        "total_tokens": 288,
      },
    },
    "tool_call_chunks": [],
    "tool_calls": [],
    "usage_metadata": {
      "input_tokens": 17,
      "output_tokens": 271,
      "total_tokens": 288,
    },
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`orchestration service client streaming supports streaming responses with tool calls 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {
      "module_results": {
        "llm": {
          "choices": [
            {
              "delta": {
                "content": "",
                "role": "assistantassistantassistantassistantassistantassistantassistantassistantassistantassistantassistant",
                "tool_calls": [
                  {
                    "function": {
                      "arguments": "{"a": 2, "b": 3}",
                      "name": "add",
                    },
                    "id": "call_HPgxxSmD2ctYfcJ3gp1JBc7i",
                    "index": 0,
                    "type": "function",
                  },
                  {
                    "function": {
                      "arguments": "{"a": 2, "b": 3}",
                      "name": "multiply",
                    },
                    "id": "call_PExve0Dd9hxD8hOk4Uhr1yhO",
                    "index": 1,
                    "type": "function",
                  },
                ],
              },
              "finish_reason": "tool_calls",
              "index": 0,
            },
          ],
          "created": 1747660479,
          "id": "chatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjdchatcmpl-BYucBnhXmLZ6QGcjM6Ac76HKLibjd",
          "model": "gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06gpt-4o-2024-08-06",
          "object": "chat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunkchat.completion.chunk",
          "system_fingerprint": "fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0fp_ee1d74bde0",
        },
        "templating": [
          {
            "content": "Add 2 and 3, and multiply 2 and 3.",
            "role": "user",
          },
        ],
      },
      "request_id": "b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05b851a041-7bcd-463f-881a-7cb40526bc05",
    },
    "content": "",
    "id": undefined,
    "invalid_tool_calls": [],
    "response_metadata": {
      "completion": 0,
      "finish_reason": "tool_calls",
      "prompt": 0,
    },
    "tool_call_chunks": [
      {
        "args": "{"a": 2, "b": 3}",
        "id": "call_HPgxxSmD2ctYfcJ3gp1JBc7i",
        "index": 0,
        "name": "add",
        "type": "tool_call_chunk",
      },
      {
        "args": "{"a": 2, "b": 3}",
        "id": "call_PExve0Dd9hxD8hOk4Uhr1yhO",
        "index": 1,
        "name": "multiply",
        "type": "tool_call_chunk",
      },
    ],
    "tool_calls": [
      {
        "args": {
          "a": 2,
          "b": 3,
        },
        "id": "call_HPgxxSmD2ctYfcJ3gp1JBc7i",
        "name": "add",
        "type": "tool_call",
      },
      {
        "args": {
          "a": 2,
          "b": 3,
        },
        "id": "call_PExve0Dd9hxD8hOk4Uhr1yhO",
        "name": "multiply",
        "type": "tool_call",
      },
    ],
    "usage_metadata": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;
